apex/contrib/sparsity/sparse_masklib.py

```diff
@@ -22,33 +23,127 @@ def reshape_1d(matrix, m):
 
 """ return all possible m:n patterns in a 1d vector """
 valid_m4n2_1d_patterns = None
+valid_m16n8_1d_patterns = None
+valid_m32n8_1d_patterns = None
+valid_m32n16_1d_patterns = None
 def compute_valid_1d_patterns(m,n):
     # Early exit if patterns was already created.
     global valid_m4n2_1d_patterns
-
+    global valid_m16n8_1d_patterns
+    global valid_m32n8_1d_patterns
+    global valid_m32n16_1d_patterns
     if m==4  and n==2 and valid_m4n2_1d_patterns  is not None: return valid_m4n2_1d_patterns
-    patterns = torch.zeros(m)
-    patterns[:n] = 1
-    valid_patterns = torch.tensor(list(set(permutations(patterns.tolist()))))
-    if m == 4 and n == 2: valid_m4n2_1d_patterns = valid_patterns
+    if m==16  and n==8 and valid_m16n8_1d_patterns  is not None: return valid_m16n8_1d_patterns
+    if m==32  and n==8 and valid_m32n8_1d_patterns  is not None: return valid_m32n8_1d_patterns
+    if m==32  and n==16 and valid_m32n16_1d_patterns  is not None: return valid_m32n16_1d_patterns
+    valid_patterns = []
+    for i in list(combinations(range(0, m), n)):
+        cur_pattern = np.zeros(m, dtype=np.int32)
+        cur_pattern[list(i)] = 1
+        valid_patterns.append(cur_pattern)
+    valid_patterns = torch.Tensor(np.array(valid_patterns))
+    # patterns = torch.zeros(m)
+    # patterns[:n] = 1
+    # valid_patterns = torch.Tensor(list(set(permutations(patterns.tolist()))))
+    if m == 4  and n == 2: valid_m4n2_1d_patterns  = valid_patterns       
+    if m == 16  and n == 8: valid_m16n8_1d_patterns  = valid_patterns       
+    if m == 32  and n == 8: valid_m32n8_1d_patterns  = valid_patterns       
+    if m == 32  and n == 16: valid_m32n16_1d_patterns  = valid_patterns       
     return valid_patterns
 
 """ m:n 1d structured best """
 def mn_1d_best(matrix, m, n):
     # Find all possible patterns.
     patterns = compute_valid_1d_patterns(m,n).cuda()
-
     # Find the best m:n pattern (sum of non-masked weights).
     mask = torch.cuda.IntTensor(matrix.shape).fill_(1).view(-1,m)
     mat,shape = reshape_1d(matrix,m)
-    pmax = torch.argmax(torch.matmul(mat.abs(),patterns.t()), dim=1)
-    mask[:] = patterns[pmax[:]]
+
+    _dynamic, m = mat.shape
+    factor = 256
+    for start in range(0, _dynamic, factor):
+        pmax = torch.argmax(torch.matmul(mat[start : start + factor].abs(), patterns.t()), dim=1)
+        mask[start: start + factor] = patterns[pmax[:]]
     mask = mask.view(matrix.shape)
+
     return mask
 
+""" m:n 1d structured pruning: greedy method to select mask """
+def mn_1d_greedy(matrix, m, n):
```

Comment:
Can you comment on why the mn_1d_greedy method is needed, and/or how its masks differ from the mn_1d_best mask for a given matrix?